nexus/src/app/sagas/instance_create.rs

```diff
@@ -758,64 +723,51 @@ async fn ensure_instance_disk_attach_state(
     attached: bool,
 ) -> Result<(), ActionError> {
     let osagactx = sagactx.user_data();
-    let disk_params = sagactx.saga_params::<DiskParams>()?;
-    let saga_params = disk_params.saga_params;
-    let disk_index = disk_params.which;
-    let opctx =
-        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
+    let disk_attachment = sagactx.saga_params::<InstanceDiskAttachment>()?;
+    let params = sagactx.saga_params::<Params>()?;
+    let datastore = osagactx.datastore();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+    let project_id = params.project_id;
 
-    let saga_disks = &saga_params.create_params.disks;
-    let instance_name =
-        db::model::Name(saga_params.create_params.identity.name);
+    let disk_name = match disk_attachment {
+        InstanceDiskAttachment::Create(create_params) => {
+            db::model::Name(create_params.identity.name)
+        }
+        InstanceDiskAttachment::Attach(attach_params) => {
+            db::model::Name(attach_params.name)
+        }
+    };
 
-    if disk_index >= saga_disks.len() {
-        return Ok(());
-    }
+    let (.., authz_instance, _db_instance) =
+        LookupPath::new(&opctx, &datastore)
+            .instance_id(instance_id)
+            .fetch()
+            .await
+            .map_err(ActionError::action_failed)?;
 
-    let disk = &saga_disks[disk_index];
+    let (.., authz_disk, _db_disk) = LookupPath::new(&opctx, &datastore)
+        .project_id(project_id)
+        .disk_name(&disk_name)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
```

Comment:
We really shouldn't be looking up disks by name here, but that's all we have. Technically I could do a special lookup for disks that were previously created and grab their IDs, but for disks that are only being attached all we have is what the API provides. 

This is a bit of an aside to this issue, but I believe that all names should be resolved inside of `http_entrypoints`. Once calls flow further down in the system the should be operating solely on IDs. Whenever we begin to implement RFD-322 we should probably consider that.

Reply:
💯 is that not worth doing here?  If not, I wonder if we should keep the comment that was here before (the TODO-security TODO-correctness one that references omicron#1536).

Reply:
I can re-add the comment. I still generally believe that we shouldn't just handle this in a one-off way, but more as a holistic solution. I know that's a non-trivial amount of work, but I still believe it to be important.